add ability to update local releases file when generating a new delta…

BryanAldrich opened the pull request:
add ability to update local releases file when generating a new delta manually or when downloading the latest full release.

Commit:
add ability to update local releases file when generating a new delta manually or when downloading the latest full release.

caesay commented:
I'm not opposed to this PR in principle but it would need some work to fix up. Adding an arg to OutputCommand is not ideal, because there are lots of commands that this argument does not apply to. Furthermore, it wouldn't be easily possible to split this into a new class, because both SomeUpload and SomeDownload share a SomeBaseCommand - and since this argument does not apply to SomeUpload, it can not be an argument shared by the base command. I guess the simplest thing would be to duplicate the argument to every SomeDownload command but that feels a bit hacky. 

Lastly, I think `-u` is used for other important arguments in other commands, so I would want to remove the short arg only leaving the long arg `--update-releases`.

Not sure I have a great idea on how I'd want to restructure things to enable this but I'll have a think about it.

Commit:
moved the implementation from the base to the download only.
Every options now has a BaseOptions if they previously had a Download and Upload. Upload no longer inherits from the Download, but from base allowing both to have differing options.
Added a Mapperly class to flip between the upload and download options, however, this only copies the base options of course. Probably shouldn't use these tbh.
